checkmate/views/api/url_check.py

```diff
@@ -0,0 +1,11 @@
+"""Monitoring views."""
+
+from pyramid import response, view
+
+
+@view.view_config(route_name="url_check")
+def url_check(_request):
+    """Status endpoint."""
+
+    # If everything is fine give a 204 which is successful, but has no body
+    return response.Response(status_int=204, status="204 OK")
```

Comment:
I think we should just be using a 200 OK here. (So `return pyramid.httpexceptions.HTTPOk`.)

204 seems to confuse Chrome into thinking the website isn't responding (possibly because of the lack of a response body) and it actually causes the Chrome tab to behave in a confused/broken way.

Who knows what it might do to other clients that're gonna be involved in production: Elastic Beanstalk and New Relic, `requests`, Cloudflare, NGINX.

It _might_ end up to work fine but it's asking for trouble.

Generally speaking trying to get semantic with HTTP status codes is unnecessary and has always gotten us into trouble:

* Monitoring tools get confused if we use 4xx's and 5xx's to mean things as they see them as errors. Possibly configurable in some monitoring tools (New Relic) but not in others (Elastic Beanstalk) and also, why bother? It's a problem entirely of our own making

* Cloudflare gets confused and injects its own HTML page if we use 5xx's

* 204 confuses Chrome

I think we should just return `200` for all URLs (whether blocked or not) and use the JSON body to indicate different results to our apps

Reply:
For a 200 OK you can also just do:

```python
@view_config(route_name="url_check")
def url_check(request):
    return request.response
```

`request.response` is a default response object that Pyramid always generates, and by default it's a normal 200 OK response.

And if you want it to be something other than a standard 200 OK you can set attrs on `request.response` to change it.

See:

* https://docs.pylonsproject.org/projects/pyramid/en/latest/whatsnew-1.1.html#request-response (particularly the second para)
* https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/advanced-features.html?highlight=request.response#use-global-response-objects

:man_shrugging: None of this matters much, I don't think it makes any real difference, it's just that you can either return HTTP exceptions or use `request.response` rather than going to the trouble of creating your own `Response` object

Reply:
I think I'll try one of the shorter ways. `request.response` feels a bit surprising to me. `HTTPOk` or `HTTPNoContent` both give you a fighting chance of googling for it an working out what's happening.

With regards to 204, I think the worry is a bit over the top:

 * I don't think Chrome is confused about 204s, rather it's reaction to them might be surprising to you
 * `curl` or a proper end-point testing tool is probably the ticket here: not all Restful interactions can be fully investigated with a browser
 * There's nothing esoteric about 204s, they are as old as HTTP pretty much, all those services are going to handle it fine
 * It's both semantically, but also operationally correct. We don't have anything to say, and this is the quickest way to not say it
 * Downside: It's going to complicate the caller slightly, as you can't blind decode the body
 * Upside: The caller doesn't have to decode a body most of the time, because there isn't one

In general I understand we've been bitten by specific cases of HTTP codes causing trouble, but I don't think we need to retreat to "200 is the only safe way". This feels ritualistic to me rather than evidence based.

Reply:
As an aside, using 204 here was my solution to avoiding the concern you had with using other codes to indicate a blocked resource. This means we aren't using error codes to let the client know.

Reply:
`request.response` is totally legit FWIW -- it's searchable in the Pyramid docs, and exists as a convenience for us to use.

I think the 204 status code itself _might_ be fine and just the empty body is what's confusing Chrome (and might confuse other things as well). But my opinion FWIW is that we should just use 200 and communicate the result to the Via's in the JSON body

Reply:
Yeah like I say, there's no confusion on Chromes part. I'm only pointing this out as I think it's part of the justification for being worried about other tools. There's just nothing to show.